refactor: favor CDN for examples

nick-romano opened the pull request:
# Description

This PR accomplishes a few items:
- [x] updates each sample and tutorial - to favor the `dist` build pattern, and use the CDN over copying assets locally
- [x] introduces syncpack to manage dependency versions for each sample
- [x] updates readmes where applicable


There's still an outstanding issue with the stackblitz link to the angular sample, which is discussed in #9 

Note: this PR is intended to be merged in the [next](https://github.com/Esri/arcgis-maps-sdk-javascript-samples-beta/tree/next) branch, which will set us up for the 4.29 release

andygup commented on packages/coding-components/templates/react/package.json:
```diff
@@ -8,20 +8,18 @@
-    "react": "18.2.0",
-    "react-dom": "18.2.0"
+    "@arcgis/coding-components-react": ">=4.29.0-beta.100 < 4.30",
+    "@arcgis/core": ">=4.29.0-next.20240207",
```
Consider just using `@next` for versioning `@arcgis/core` for the various samples.

andygup commented on packages/coding-components/templates/react/package.json:
```diff
@@ -8,20 +8,18 @@
     "typescript": "~5.3.0",
-    "vite": "^5.0.12",
-    "vite-plugin-static-copy": "^1.0.1"
+    "vite": "^5.0.12"
```
Maybe run `ncu` again? Vite latest is 5.1.x for the various samples.

andygup commented on packages/coding-components/templates/webpack/package.json:
```diff
@@ -6,15 +6,13 @@
     "html-webpack-plugin": "^5.5.0",
-    "resolve-pkg": "^2.0.0",
     "style-loader": "^3.3.1",
     "webpack": "^5.84.0",
```
Recheck these deps, e.g. `ncu`. webpack latest is 5.90.1

andygup commented on packages/coding-components/templates/webpack/readme.md:
```diff
@@ -24,57 +24,21 @@ yarn run start
-
 #### JS
 
 We imported the components we need for the coding components by following [Stencil's instructions for integrating components without a JavaScript framework](https://stenciljs.com/docs/javascript).
```
```suggestion
Import the components using [Stencil's instructions for integrating components without a JavaScript framework](https://stenciljs.com/docs/javascript).
```

andygup commented on packages/coding-components/templates/webpack/readme.md:
````diff
@@ -24,57 +24,21 @@ yarn run start
+defineCodingElements(window, { resourcesUrl: "https://js.arcgis.com/coding-components/next/assets" });
 ```
 
 We use [`src/index.js`](./src/index.js) to load our data, define our custom elements, and utilize various kinds of properties in the editor. It is a must to define both the Calcite and Coding elements.
````
```suggestion
Use [`src/index.js`](./src/index.js) to load the data, define custom elements, and utilize various properties in the editor. Both the Calcite and Coding elements must be defined.
```

andygup commented on packages/map-components/templates/react/src/index.tsx:
```diff
@@ -2,6 +2,10 @@ import React from "react";
+// Register custom elements
+defineCustomElements();
 
 import "@arcgis/core/assets/esri/themes/dark/main.css";
```
Hmmm, double check the CSS. It might need to be the CDN and not pulled from node_modules.

nick-romano replied:
Works OK

andygup replied:
Yep it will work, but for consistency the CSS should be using CDN here.

andygup replied:
Sorry, I should have given a link. Here's the pattern we've been using for ESM samples: https://github.com/Esri/jsapi-resources/blob/main/esm-samples/jsapi-react/src/index.css#L1

```css
@import 'https://js.arcgis.com/<version>/@arcgis/core/assets/esri/themes/light/main.css';
```

andygup commented on packages/map-components/templates/vite/package.json:
```diff
@@ -8,9 +8,9 @@
+    "@esri/calcite-components": "^2.4.0"
   },
   "devDependencies": {
     "vite": "^5.0.12"
```
Check dep for @latest, e.g. `ncu`. Latest Vite is 5.1.x.

andygup commented on packages/map-components/templates/vue/package.json:
```diff
@@ -8,9 +8,9 @@
+    "@arcgis/core": ">=4.29.0-next.20240207",
+    "@arcgis/map-components": ">=4.29.0-beta.100 < 4.30",
+    "@esri/calcite-components": "^2.4.0",
     "vue": "^3.3.4"
```
Same, check `vue` dep for @latest.

andygup commented on packages/map-components/templates/webpack/package.json:
```diff
@@ -6,9 +6,9 @@
+    "@esri/calcite-components": "^2.4.0"
   },
   "devDependencies": {
     "css-loader": "^6.7.2",
```
Check dep for latest, I think css-loader is 6.10.0. Some of these libraries sneak in breaking changes, etc in minor releases that should always be verified.

andygup commented on packages/map-components/tutorials/create-a-web-app-using-components-solution/package.json:
```diff
@@ -8,9 +8,9 @@
+    "@esri/calcite-components": "^2.4.0"
   },
   "devDependencies": {
     "vite": "^5.0.12"
```
Check dep for @latest.

andygup reviewed (changes requested):
Thanks for kicking this off Nick! There were a number of dependencies that need verification and updating. Also, once all the deps are updated, I recommend locally testing (or retesting) all samples by running a dist build. It's well worth the effort to catch things early if it uncovers an SDK bug, or a dependency bug, or ...

nick-romano commented:
OK I think all the requested changes have been applied. I checked all the samples last night and they work well locally. 

- [x] updated deps like webpack, vue, vite, etc. to latest versions
- [x] made sure each map component sample had an example of listening for `onArcgisViewReady`
- [x] Changed around the vue sample to add an event callback
- [x] updated the angular pattern to use a provider

Ready for another 👁️ @andygup @omarkawach

omarkawach commented on packages/coding-components/templates/angular/src/app/app.module.ts:
```diff
@@ -13,20 +13,35 @@
+import { APP_INITIALIZER, NgModule } from "@angular/core";
+import { BrowserModule } from "@angular/platform-browser";
+// Must be included for component packages that don't yet have an Angular wrapper
+import { CUSTOM_ELEMENTS_SCHEMA } from "@angular/core";
```
I originally included this because Calcite didn't have an Angular wrapper at the time. This should no longer be necessary if we include [ `@esri/calcite-components-angular`](https://www.npmjs.com/package/@esri/calcite-components-angular) in our `package.json`

omarkawach commented on packages/coding-components/templates/webpack/webpack.config.js:
```diff
@@ -14,8 +14,6 @@
  */
 
 const HtmlWebPackPlugin = require("html-webpack-plugin");
```
While we're removing plugins, it may be time to remove this plugin as well. WDYT?

andygup replied:
Good point. I need to look into doing the same thing for the jsapi-resources webpack sample.

nick-romano replied:
update: decision was made to keep the html plugin around.

omarkawach commented on packages/map-components/templates/react/src/index.tsx:
```diff
@@ -1,14 +1,22 @@
+    <ArcgisMap 
+      item-id="d5dda743788a4b0688fe48f43ae7beb9" 
+      onArcgisViewReadyChange={(event: any) => {
+      console.log('Map View ready', event);
```
I might've documented this incorrectly somewhere. It should be `MapView` instead of `Map View`

odoe replied:
React samples should use `itemId` instead of `item-id`.

omarkawach commented on packages/map-components/templates/vue/src/App.vue:
```diff
@@ -1,6 +1,36 @@
+  console.log("Map View ready", event);
+}
+
+const popupOptions = defineProps({
+  dockEnabled: {
+    type: Boolean,
+    default: () => true
+  },
+  dockOptions: {
+    type: Object,
+    default: () => {
+      return {
+        position: 'bottom-right',
+        breakpoint: false
+      }
+    }
+  }
+})
```
Great example on how to set properties on a map component in Vue! However, if we're doing this in one template then it might be best to do this in all templates for consistency?

nick-romano replied:
removed it for now

omarkawach reviewed (changes requested):
Looking good! Only a few comments

andygup commented on .syncpackrc.js:
```diff
@@ -0,0 +1,132 @@
+/* Copyright 2023 Esri
```
```suggestion
/* Copyright 2024 Esri
```

andygup commented on packages/coding-components/templates/amd-script-tag/index.html:
```diff
@@ -29,17 +29,17 @@
       rel="stylesheet"
       type="text/css"
-      href="https://js.arcgis.com/coding-components/4.28/arcgis-coding-components.css"
+      href="https://js.arcgis.com/coding-components/next/arcgis-coding-components.css"
```
Omar, is components CDN-next is sync with the AMD CDN-next releases? Just double checking.

omarkawach replied:
It's in sync on dev. 

js.arcgis.com for components should be in sync again end of February.

andygup commented on packages/coding-components/templates/angular/angular.json:
```diff
@@ -18,7 +18,7 @@
             "main": "src/main.ts",
             "polyfills": ["zone.js"],
```
I'm adding the zone.js monkey patch hack to the jsapi-resources Angular sample at 4.29: https://github.com/Esri/jsapi-resources/tree/main/esm-samples/jsapi-angular-cli#known-issues and here's the commit: https://github.com/Esri/jsapi-resources/commit/43ba834d95905a50301ce505ac1bf38b4b953ce3. 

The reason for this is a number of API widgets are broken due to a known bug in Calcite including zoom and editor. And, we didn't want to publish a sample that didn't work with two of the most popular widgets.

nick-romano replied:
So does it feel appropriate to add it here? seems like a yes to me?

andygup commented on packages/coding-components/templates/angular/package.json:
```diff
@@ -1,29 +1,25 @@
+  "private": true,
   "scripts": {
     "ng": "ng",
     "start": "ng serve -o",
```
Consider switching out with ` "start": "ng serve --force-esbuild",` it's much faster 😄

andygup commented on packages/coding-components/templates/angular/package.json:
```diff
@@ -1,29 +1,25 @@
+    "@angular/platform-browser-dynamic": "^17.1.0",
+    "@angular/router": "^17.1.0",
+    "@arcgis/coding-components-angular": "next",
+    "@arcgis/core": "rc",
```
@arcgis/core should be `next` or a production version. `rc` tagged versions only exist for a few patches before the production release.

nick-romano replied:
I should've outlined this in the PR description, but the plan is that:

1. we merge this branch into the `next` branch
2. the day of/before the release (which is better?), we open a pr from next into main, and in that PR, we change all `next` AND `rc` pointers to `4.29`. Then main is set with all the samples relevant to our latest release.
3. we delete `next`
4. eventually we create a new `next` branch in the following release cycle

and while I type that out, maybe we shouldn't call it the `next` branch, but instead `4.29-release`?

andygup replied:
Oh, gotcha, thanks for the background info, Nick. Do we need that granular of version control for the public samples e.g. next, rc and latest? Most customers are going to be using `latest` for their production apps, and if someone wants to test with another version it's easy enough for them to npm install. I also see that `@arcgis/coding-components-angular` doesn't have a corresponding `rc` tag on npm. Could be confusing.

andygup commented on packages/coding-components/templates/angular/src/app/app.module.ts:
```diff
@@ -13,20 +13,35 @@
+  providers: [
+    {
+      provide: APP_INITIALIZER,
+      useFactory: () => initializeCustomElements,
```
Just a heads-up, this is likely to trip up customers who don't understand factory providers or can't use this approach due to their existing application architecture. Often when that happens they'll call support or say it's a bug or defect in the SDK. Is there a more generic way to initialize the component library?

nick-romano replied:
unfortunately there is not. The only alternative would be to include `schema: [CUSTOM_ELEMENTS_SCHEMA]`, and removing the wrappers from the root `NgModule`. FWIW I spent a lot of time on this Monday and came to the conclusion that this is the cleanest and most angular compliant approach. I'll send you a doc on the subject.

andygup replied:
I chatted with Nick about investigating using `ngOnit()` which is a standard Angular component life-cycle event like mapViewReady.

andygup commented on packages/coding-components/templates/vite/package.json:
```diff
@@ -8,13 +8,11 @@
-    "@arcgis/core": "~4.28.0",
-    "@esri/calcite-components": "^1.9.2"
+    "@arcgis/coding-components": "next",
+    "@arcgis/core": "rc",
```
Ditto on `rc` tag

andygup reviewed (changes requested):
Mostly small stuff and minor cleanup. The `@arcgis/core` package versions need to be adjusted. Thanks!

nick-romano commented on packages/map-components/templates/angular/package.json:
```diff
@@ -1,38 +1,38 @@
-    "@angular/cli": "~16.2.1",
-    "@angular/compiler-cli": "^16.2.0",
+    "@angular-devkit/build-angular": "^17.0.0",
+    "@angular/cli": "^17.0.0",
```
ahh good find!

nick-romano commented on packages/map-components/templates/react/package.json:
```diff
@@ -8,18 +8,18 @@
     "@vitejs/plugin-react": "^4.2.1",
     "typescript": "~5.3.0",
-    "vite": "^5.0.12"
+    "vite": "5.1.1"
```
would `^5.1.1` be better? since these are samples I kind of like having dev deps locked down, but patches and minor update *shouldn't* break anything

nick-romano commented on packages/map-components/tutorials/create-a-web-app-using-components-solution/package.json:
```diff
@@ -8,11 +8,11 @@
   },
   "devDependencies": {
-    "vite": "^5.0.12"
+    "vite": "5.1.1"
```
just fyi, I introduced syncpack [here](https://github.com/Esri/arcgis-maps-sdk-javascript-samples-beta/blob/nr/use-cdn/.syncpackrc.js) to sync all the deps for each sample. So that's why you're seeing the same `5.1.1` or `rc` in each project, they are pinned in syncpack and then reflected in each package.json. Open to comments there on what should be the proper versions

andygup replied:
We've been going with the dependencies and versioning that the framework installs, as their default. That [semantic versioning](https://docs.npmjs.com/about-semantic-versioning) will be what the framework themselves supports. Also it helps us tell a story that we are always striving to be compatible with the "latest release", for example, "Vite 5". And if there's an issue with the framework, bundler, component or @arcgis/core the typical best practice is to upgrade to the latest to get any potential fixes that have been (or will be) released.

nick-romano replied:
makes sense!

nick-romano commented on packages/coding-components/templates/vite/vite.config.js:
```diff
@@ -13,31 +13,10 @@
-    }),
-  ],
+  plugins: [],
   build: {
```
@omarkawach - should we do this for all the samples?

omarkawach replied:
Ya let's do it for consistency

nick-romano commented on packages/coding-components/templates/react/package.json:
```diff
@@ -8,20 +8,18 @@
-    "react": "18.2.0",
-    "react-dom": "18.2.0"
+    "@arcgis/coding-components-react": "next",
+    "@arcgis/core": "rc",
```
they are until core starts publishing `rc` versions and points `next` to the following release (4.30 currently).

bsvensson replied:
core has been publishing `rc` versions (and 4.30) for two weeks already.  
https://www.npmjs.com/package/@arcgis/core?activeTab=versions

omarkawach reviewed (approved):
Looks good to me. Let's merge this into `next` and create another PR for a few small changes:
- Update root README.md file with get started instructions for users who might git clone this yarn workspaces monorepo
- Fix angular tests for map-components, and potentially any other components

andygup reviewed (approved):
Other than the window.location.href I don't have any other immediate changes. I am still a bit concerned about the installation complexity of the repo and users natural desire to clone for a variety of well known GitHub reasons. We'll see what feedback we get.

nick-romano commented:
> Other than the window.location.href I don't have any other immediate changes. I am still a bit concerned about the installation complexity of the repo and users natural desire to clone for a variety of well known GitHub reasons. We'll see what feedback we get.

@andygup - could you elaborate on the `window.location.href` change? I think it might be lost in the size of this PR.

andygup commented:
Good point @nick-romano, too many repos and changes in one day. The window.location.href comment was in regards to Components gettings started with React. But, I wasn't sure if that code made its way into this repo.
